plugins/quantum_k_nearest_neighbours/backend/qknns/schuld_hamming.py

```diff
@@ -0,0 +1,169 @@
+    max_num_decimal_places = [0] * X.shape[1]
+    for vec in X:
+        for dim in range(X.shape[1]):
+            max_num_decimal_places[dim] = max(
+                num_decimal_places(str(vec[dim])), max_num_decimal_places[dim]
+            )
+    max_num_bits = [0] * X.shape[1]
+    for vec in X:
+        for dim in range(X.shape[1]):
+            temp = vec[dim] * (10 ** max_num_decimal_places[dim])
+            if temp < 0:
+                temp *= -1
+            temp = int(np.ceil(np.log2(temp)))
+            max_num_bits[dim] = max(temp, max_num_bits[dim])
+    result = []
+    for vec in X:
+        new_vec = []
+        for dim in range(X.shape[1]):
+            new_vec += int_to_bitlist(
+                int(vec[dim] * max_num_decimal_places[dim]), max_num_bits[dim]
+            )
+        result.append(new_vec)
+    return np.array(result)
+
+
+class SchuldQkNN(QkNN):
+    def __init__(
+        self,
+        train_data,
+        train_labels,
+        train_wires: List[int],
+        label_wires: List[int],
+        qam_ancilla_wires: List[int],
+        backend: qml.Device,
+        unclean_wires=None,
+    ):
+        super(SchuldQkNN, self).__init__(
+            train_data, train_labels, len(train_data), backend
+        )
+        self.train_data = np.array(train_data, dtype=int)
+
+        if not check_if_values_are_binary(self.train_data):
+            raise ValueError(
+                "All the data needs to be binary, when dealing with the hamming distance"
+            )
+
+        self.label_indices = self.init_labels(train_labels)
+
+        self.unclean_wires = [] if unclean_wires is None else unclean_wires
+
+        self.train_wires = train_wires
+        self.qam_ancilla_wires = qam_ancilla_wires
+        self.label_wires = label_wires
+        wire_types = ["train", "label", "qam_ancilla", "unclean"]
+        num_wires = [
+            self.train_data.shape[1],
+            self.label_indices.shape[1],
+            max(self.train_data.shape[1], 2),
+        ]
+        error_msgs = [
+            "the points' dimensionality.",
+            "ceil(log2(len(unique labels))).",
+            "the points' dimensionality and greater or equal to 2.",
+        ]
+        check_wires_uniqueness(self, wire_types)
+        check_num_wires(self, wire_types[:-1], num_wires, error_msgs)
+
+        self.qam = QAM(
+            self.train_data,
+            self.train_wires,
+            self.qam_ancilla_wires,
+            additional_bits=self.label_indices,
+            additional_wires=self.label_wires,
+            unclean_wires=unclean_wires,
+        )
+
+    def init_labels(self, labels):
+        label_indices = list()
+        label_to_idx = (
+            dict()
+        )  # Map labels to their index. The index is represented by a list of its bits
+        num_bits_needed = int(
+            np.ceil(np.log2(len(self.unique_labels)))
+        )  # Number of bits needed to represent all indices of our labels
```

Comment:
I have often seen this combination of `int(np.ceil(np.log2(*)))`. This could be refactored into a utility function for clarity giving a function name to this composed function.